fleetspeak/src/client/daemonservice/execution/execution.go

```diff
@@ -363,42 +382,38 @@ func (e *Execution) Shutdown() {
 	})
 }
 
-// inLoop reads messages from the dependent process and passes them to
+// inRoutine reads messages from the dependent process and passes them to
 // fleetspeak.
-func (e *Execution) inLoop() {
+func (e *Execution) inRoutine() {
 	defer func() {
 		e.Shutdown()
 		e.inProcess.Done()
 	}()
-	initialMsg := e.readMsg()
-	if initialMsg == nil {
-		return
-	}
 
-	if initialMsg.Destination != nil && initialMsg.Destination.ServiceName == "system" && initialMsg.MessageType == "StartupData" {
-		sd := &fcpb.StartupData{}
-		if err := ptypes.UnmarshalAny(initialMsg.Data, sd); err != nil {
-			log.Warningf("Failed to parse startup data from initial message: %v", err)
-		} else {
-			e.startupData <- sd
-			close(e.startupData) // No more values to send through the channel.
-		}
-	} else {
-		// Handle initial message like we would any other.
-		e.setLastActive(time.Now())
-		if err := e.sc.Send(context.Background(), service.AckMessage{M: initialMsg}); err != nil {
-			log.Errorf("error sending message to server: %v", err)
-		}
-	}
 	for {
 		m := e.readMsg()
 		if m == nil {
 			return
 		}
 		e.setLastActive(time.Now())
-		if err := e.sc.Send(context.Background(), service.AckMessage{M: m}); err != nil {
-			log.Errorf("error sending message to server: %v", err)
+		e.recordHeartbeat()
+
+		if m.Destination != nil && m.Destination.ServiceName == "system" && m.MessageType == "StartupData" {
```

Comment:
This logic block is awkward to read. How about:
if m.Destination != nil && m.Destination.ServiceName == "system" {
  switch m.MessageType {
    case "StartupData":
    ...
   case "Heartbeat":
   default:
     log.Warning("Unknown system message type: %s", ...)
  }
 continue
}
Send(...)

Reply:
Done.